Code/GraphMol/FindRings.cpp

```diff
@@ -1098,112 +1096,112 @@ int symmetrizeSSSR(ROMol &mol) {
+  RingUtils::convertToBonds(extras, bondextras, mol);
+
+  //
+  // For each "extra" ring, figure out if it could replace a single
+  // ring in the SSSR. A ring could be swapped out if:
+  //
+  // * They are the same size
+  // * The replacement doesn't remove any bonds from the union of the bonds
+  //   in the SSSR.
+  //
+  // The trick is that we count how many times each bond is covered by the
+  // existing SSSR, and then decrement that count for the removed ring and
+  // increment it for the added ring.
+  //
+  // May miss extra rings that would need to swap two (or three...) rings
+  // to be included.
+
+  // counts of each bond
+  std::unordered_map<int, int> bondCounts;
+  for (const auto& r: bondsssrs) {
+    for (const auto& b: r) {
+      bondCounts[b] += 1;
+    }
+  }
+
+  std::unordered_map<int, int> activeBondCounts;
+  auto extraAtomRing = extras.begin();
+  for (auto& extraBondRing: bondextras) {
+    bool foundAMatch = false;
+    for (auto& ring: bondsssrs) {
+      if (ring.size() != extraBondRing.size()) {
+        continue;
+      } else if (!shareABond(extraBondRing, ring)) {
         continue;
       }
-      exr = bextra[eid];
-#if 0
-          std::cerr<<"      "<<eid<<": ";
-          std::copy(extras[eid].begin(),extras[eid].end(),std::ostream_iterator<int>(std::cerr," "));
-          std::cerr<<"  | ";
-          std::copy(exr.begin(),exr.end(),std::ostream_iterator<int>(std::cerr," "));
-          std::cerr<<std::endl;
-#endif
-      if (ssiz == exr.size()) {
-        // std::cerr<<"        possible"<<std::endl;
-        INT_VECT eunion;
-        Union(nunion, exr, eunion);
-#if 0
-            std::cerr<<"           munion: ";
-            std::copy(munion.begin(),munion.end(),std::ostream_iterator<int>(std::cerr," "));
-            std::cerr<<std::endl;
-            std::cerr<<"           eunion: ";
-            std::copy(eunion.begin(),eunion.end(),std::ostream_iterator<int>(std::cerr," "));
-            std::cerr<<std::endl;
-#endif
-        // now check if the eunion is same as the original union from the SSSRs
-        if (eunion.size() == munion.size()) {
-          // we found a symmetric ring
-          symids.push_back(eid);
-          // std::cerr<<"        keep!"<<std::endl;
+
+      // Remove the SSSR ring bonds and add the
+      // extra ring bonds. If any bonds are missing (count = 0),
+      // the extra ring is not a replacement.
+      activeBondCounts.clear();
+      for (auto& bondID: ring) {
+        activeBondCounts[bondID] = bondCounts[bondID] - 1;
+      }
+
+      for (auto& bondID: extraBondRing) {
+        activeBondCounts[bondID] += 1;
+      }
+
+      bool foundAMatch = true;
+      for (auto& bondIDAndCount: activeBondCounts) {
+        if (bondIDAndCount.second < 1) {
+          foundAMatch = false;
+          break;
         }
       }
+
+      if (foundAMatch) {
+        res.push_back(*extraAtomRing);
+        FindRings::storeRingInfo(mol, *extraAtomRing);
+        break;
```

Comment:
If you get to this point, don't you need to update the boundCounts?

I mean, you don't want to include other `extraRings` in `res` that replace the same bonds that `extraAtomRing`, while you forget about the ones that you haven't replaced yet (the ones that were in `ring` but not in `extraBondRing`).

Reply:
No, because we don't update the SSSR that we base this on. This always searches the original SSSR for a ring that can be swapped.

I think that updating `bondCounts` would be wrong, because then we'd be in danger of finding rings that actually need two rings to replace the original ring.

Reply:
I see. But I was thinking whether it was possible that, through successive swaps, at some point any of the bonds in the SSSR was no longer "covered".

I.e., imagine that one specific `bond` is common to 2 rings (`ring1` and `ring2`), both of them in the SSSR. So, this bond would have `bondCount` = 2.

In one iteration, `ring1` gets swapped with `extraRing1`, which has some `otherBond1` in common with `ring1`, but does not include `bond`. That's ok, because `ring1` is not the only provider of `bond` in the SSSR.

In a latter iteration, `ring2` gets swapped with `extraRing2`, which has some `otherBond2` in common with `ring2`, but does neither include `bond`. That's also ok, because `ring2` is neither the only provider of `bond` in the SSSR.

But, then, in `res` none of the bonds would provide coverage for `bond`, since `ring1` and `ring2`, the only providers of `bond` have been swapped out.

(Is such a situation even possible?)

Reply:
But we don't swap out the bonds in the SSSR. We're adding to the ring set, not replacing anything.

This is a sneaky way to find all rings that are in any SSSR.